lib/Ast.ml

```diff
@@ -2228,6 +2228,14 @@ end = struct
         false
     (* Object fields do not require parens, even with trailing attributes *)
     | Exp {pexp_desc= Pexp_object _; _}, _ -> false
+    | ( Exp
+          { pexp_desc=
+              Pexp_sequence
+                (({pexp_desc= Pexp_match _ | Pexp_try _; _} as e), _)
+          ; _ }
+      , {pexp_attributes= _ :: _; _} )
```

Comment:
Could you add a comment explaining this case ?
It is quite confusing, especially the `pexp_attributes` pattern.

This is a special case of the next pattern, which seems to be a special case of the following pattern and `Pexp_sequence` near the end.
This is quite hard to understand. What about moving everything related to sequence to an other function and keeping only one `Exp {pexp_desc = Pexp_sequence}` here.
There is only 2 generic patterns and they would need to be duplicated (I think this is a win in readability, especially since there is so much special handling): `Exp.has_trailing_attributes` and `_` in `Exp` near the end.

Reply:
I improved the situation a bit, but the whole mechanism needs to be rewritten some day, it's painful to maintain.